File: 5_Basic_Recursion/5Factorial.cpp
#include <iostream>
using namespace std; 

/*  Fctorial of N  =  1*2*3...*N    

1.  PARAMETERIZED WAY =   */


void fac(int n, int prod =1){ 
  
    if (n<1){
       cout << prod ;
       return; 
    }
    prod *= n;
    n--;
    fac(n,prod);
}

/* 2.  FUNCTIONAL WAY =    */
int factorial(int n){
    if(n==0){
        return 1;
    }
    return n* factorial(n-1);

}



int main(){
    int n;
    cin >> n;
    cout << factorial(n);
}

File: 5_Basic_Recursion/5Factorial.py
#  Factorial of N  =  1*2*3...*N    

#   1.  FUNCTIONAL WAY =   

def fac(n):
    if(n==0):
        return 1
    return n* fac(n-1)


print(fac(6))


#   2.   PARAMETERIZED WAY  = 
def factorial(n,prod=1):
    if(n==0):
        print(prod)
        return 1
                                #   prod = prod * n             
    factorial(n-1, prod*n)      #   n =n-1 
                                #   factorial(n, prod)

factorial(5)
